[indexer] Compatibility check using migration records

lxfind opened the pull request:
## Description 

Previously we check DB compatibility by making sure that we could make select query on all columns to the DB based on the locally known schema. This doesn't cover all cases, for instance, there could be tables in the DB that does not exist in the local schema.
This PR changes how we do the compatibility check by fully leveraging the migration records. It checks that the migration records in the DB must fully match with the locally known one.
It also moves the code to sui-indexer crate, so that we could do this check on the startup of both indexer and graphql server.
This does require that from now on we fully respect the migration scripts, and don't do adhoc modifications on the existing migration.

## Test plan 

CI

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

gegaowp commented on crates/sui-indexer/src/db.rs:
```diff
@@ -128,13 +135,63 @@ pub fn get_pool_connection(pool: &ConnectionPool) -> Result<PoolConnection, Inde
+    );
+    for i in 0..max(known_migrations.len(), db_migrations.len()) {
+        let local_migration_record = known_migrations.get(i).cloned().unwrap_or_default();
+        let db_migration_record = db_migrations.get(i).cloned().unwrap_or_default();
```
when a new migration is added in a pr, `local_migration_record` will have an incremental migration, when is the new migration going to be applied to `db_migration_record`?

lxfind replied:
My understanding is that as part of our release process, someone would start a new indexer instance that contains the new schema with the new migration scripts and backfill it when we want to release a new indexer with schema changes.

gegaowp commented on crates/sui-indexer/src/db.rs:
```diff
@@ -128,13 +135,63 @@ pub fn get_pool_connection(pool: &ConnectionPool) -> Result<PoolConnection, Inde
+    pub version: String,
+}
+
+pub async fn check_db_migration_consistency(pool: ConnectionPool) -> Result<(), IndexerError> {
```
I saw this pr is targeting `main`, shall we base it on `idx-breaking-change-park` as we have some `migrations` changes there, which will be problematic to merge if this pr merges first.

lxfind replied:
Could you elaborate what the problem is if we merge this to main today? Did we introduce migrations to main without a parking branch?

amnn commented on crates/sui-graphql-rpc/src/server/builder.rs:
```diff
@@ -88,10 +88,9 @@ impl Server {
-        check_all_tables(&self.db_reader).await?;
-        info!("Compatibility check passed");
+        // Make sure to run all migrations on startup, and also serve as a compatibility check.
+        run_migrations(&mut get_pool_connection(&self.db_reader.inner.get_pool())?)
```
I don't think GraphQL (or any reader) should run migrations directly -- just check that they have run. GraphQL's connection to the DB should not generally have write permissions, and I think that's a good property to preserve to avoid mistakes.

bmwill replied:
Agree with this. We should even think about giving our readers (graphql, jsonrpc) read-only user roles on the DB so that they cannont write to the db

amnn commented on crates/sui-indexer/src/main.rs:
```diff
@@ -50,6 +53,9 @@ async fn main() -> Result<(), IndexerError> {
         }
         Command::JsonRpcService(json_rpc_config) => {
+            // Make sure to run all migrations on startup, and also serve as a compatibility check.
+            run_migrations(&mut get_pool_connection(&connection_pool)?)?;
```
Ditto -- readers shouldn't run migrations.

amnn reviewed (commented):
I think we do still need an asymmetry between the read side (which checks the migrations it's aware of are a prefix of the migrations that have been run on the DB), and the write side.

In general, I would be a fan of decoupling running the migrations from checking that they have run. I think based on our latest discussion about the release process, that's the direction we are headed in right? So that means both the reader and writer will check for migration compatibility (the prefix check that I mentioned in the inline comment), and we will run migrations we need to against the DB as part of the release process.

lxfind commented:
Addressed feedback.
The migration now only run on the writer startup path.
For reader we only check for consistency.
Also added a dedicated command to run migrations in case we need to do so without starting up a writer.

amnn commented on crates/sui-indexer/src/db.rs:
```diff
@@ -122,16 +129,68 @@ pub fn reset_database(conn: &mut PoolConnection) -> Result<(), anyhow::Error> {
+    .into_iter()
+    .map(|m| m.version)
+    .collect();
+    db_migrations.sort();
```
nit: Is this necessary? There's an `ORDER BY` on the query.

amnn commented on crates/sui-indexer/src/db.rs:
```diff
@@ -122,16 +129,68 @@ pub fn reset_database(conn: &mut PoolConnection) -> Result<(), anyhow::Error> {
+        "Migration Records from local schema: {:?}",
+        known_migrations
+    );
+    for i in 0..db_migrations.len() {
+        let local_migration_record = known_migrations.get(i).cloned().unwrap_or_default();
```
Check that `known_migrations` is shorter than `db_migrations` and then zip them together to avoid the unwraps?

amnn replied:
Today this looks like it's checking that every migration in the DB is known in the embedded migrations.  Shouldn't it be the other way around? (All the embedded migrations should be in the DB) to support an older version of the software running on a DB that has had backwards compatible migrations run on it?

lxfind replied:
This is tricky.
So one potential problem with "All the embedded migrations should be in the DB", would be that once a reader binary contains new migration scripts, we must update writer (or manually run migrations) before we could update the reader.
Maybe what we should do is to check that either one is the prefix of another, but never diverge?
Reads are supposed to be compatible either way.

lxfind replied:
In anyway, I updated that part (reverted the prefix comparison), and added tests.

amnn commented on crates/sui-indexer/src/db.rs:
```diff
@@ -185,9 +244,31 @@ pub mod setup_postgres {
+    // Check that the migration records in the database created from the local schema
+    // pass the consistency check.
+    #[tokio::test]
+    async fn db_migration_consistency_smoke_test() {
```
Can we add a test for the backwards compatible case? (Strict prefix)

gegaowp commented on crates/sui-indexer/src/db.rs:
```diff
@@ -122,16 +136,67 @@ pub fn reset_database(conn: &mut PoolConnection) -> Result<(), anyhow::Error> {
+pub fn run_migrations(conn: &mut PoolConnection) -> Result<(), anyhow::Error> {
+    conn.run_pending_migrations(MIGRATIONS)
+        .map_err(|e| anyhow!("Failed to run migrations {e}"))?;
+    info!("Ran migrations.");
```
nit: useful to have the latest version in the `info!`.

gegaowp commented on crates/sui-indexer/src/db.rs:
```diff
@@ -175,19 +240,74 @@ pub mod setup_postgres {
+        // Local migrations is one record more than the applied migrations.
+        // This will fail the consistency check since it's not a prefix.
+        assert!(check_db_migration_consistency(&mut conn).is_err());
+    }
```
test runing `run_migrations` then assert it will pass again?

gegaowp commented on crates/sui-indexer/src/db.rs:
```diff
@@ -122,16 +136,67 @@ pub fn reset_database(conn: &mut PoolConnection) -> Result<(), anyhow::Error> {
+/// existing readers fail over.
+/// We do however need to make sure that whenever we are deploying a new version of either reader or writer,
+/// we must first run migration scripts to ensure that there is not more local scripts than in the DB record.
+pub fn check_db_migration_consistency(conn: &mut PoolConnection) -> Result<(), IndexerError> {
```
iiuc the prefix check is about, a pod can continue working when the binary local schema is behind the DB applied schema, is it expected in our release cycle? assuming that we will deploy indexer writer, reader and graphql following the same cadence.

gegaowp reviewed (approved):
great move!
